File: C++/710.cpp
__________________________________________________________________________________________________
sample 164 ms submission
static int _ = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    return 0;
} ();
class Solution {
    int N;
    vector<int> B;
public:
    Solution(int N, vector<int> blacklist): N(N-blacklist.size()), B(move(blacklist)) {
        sort(B.begin(), B.end());
        int i=0;
        for (int &b: B) {
            b -= (i++);
        }
    }
    
    int pick() {
        int n = rand()%N;
        if (B.empty() || n<B[0]) return n;
        int l = 0, r=B.size()-1;
        while (l<r) {
            int mid = (l+r)/2;
            if (B[mid]<=n && n<B[mid+1]) return n+mid+1;
            if (B[mid]>n) r=mid-1; else l=mid+1;
        }
        return n+l+1;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(N, blacklist);
 * int param_1 = obj.pick();
 */
__________________________________________________________________________________________________
sample 50128 kb submission
class Solution {
public:
    vector<int> blacklist;
    int sz;
    Solution(int N, vector<int> blacklist) {
        this->blacklist = blacklist;
        sort(this->blacklist.begin(), this->blacklist.end());
        sz = N - this->blacklist.size();
    }
    
    int pick() {
        int r = rand() % sz;
        int lo = 0;
        int hi = this->blacklist.size();
        while (lo < hi) {
            int md = (lo + hi) / 2;
            if (this->blacklist[md] >= r + md + 1) {
                hi = md;
            } else {
                lo = md + 1;
            }
        }
        return r + lo;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(N, blacklist);
 * int param_1 = obj.pick();
 */
__________________________________________________________________________________________________

File: Python3/710.py
__________________________________________________________________________________________________
sample 320 ms submission  
import math
import random
class Solution:

    def __init__(self, N: int, blacklist: List[int]):
        b_set = set(blacklist)
        self.N = N
        self.r = r = N - len(blacklist)
        upper_whitelist = [i + self.r for i in range(len(blacklist)) if (i + self.r) not in b_set]
        lower_blacklist = [b for b in blacklist if b < r]
        self.map = {}
        for i, b in enumerate(lower_blacklist):
            self.map[b] = upper_whitelist[i]
    
    def pick(self) -> int:
        rand = math.floor(random.random() * self.r)
        if rand not in self.map:
            return rand
        return self.map[rand]
        


# Your Solution object will be instantiated and called as such:
# obj = Solution(N, blacklist)
# param_1 = obj.pick()
__________________________________________________________________________________________________
sample 21992 kb submission
class Solution:

    def __init__(self, N: int, blacklist: List[int]):
        blacklist = set(blacklist)
        self.length = N - len(blacklist)
        self.remap = {}
        need_remap = []
        
        for i in blacklist:
            # 这里的i既在blacklist中，同时也是会被random函数选的潜在数值。所以要remap
            # 到whitelist中的某个数.
            # i < self.length的意思是length表示whitelist的长度，i 比length小是可能被选到的
            # 所以要加到需要被remap的名单里
            if i < self.length:
                need_remap.append(i)
                
        j = 0
        
        for i in range(self.length, N):
            if i not in blacklist:
                self.remap[need_remap[j]] = i
                j += 1

    def pick(self) -> int:
        idx = random.randrange(self.length)
        return self.remap[idx] if idx in self.remap else idx

# Your Solution object will be instantiated and called as such:
# obj = Solution(N, blacklist)
# param_1 = obj.pick()
__________________________________________________________________________________________________
